File: apps/api/internal/domain/eta.go
package domain

import "math"

// EstimateETAMinutes returns an approximate driving ETA using haversine distance
// at avgSpeedKmh (defaults to 40 km/h when <= 0). Returns 0 when coords are invalid.
func EstimateETAMinutes(fromLat, fromLng, toLat, toLng, avgSpeedKmh float64) int {
	if fromLat == 0 && fromLng == 0 {
		return 0
	}
	if toLat == 0 && toLng == 0 {
		return 0
	}
	if avgSpeedKmh <= 0 {
		avgSpeedKmh = 40
	}
	km := haversineKm(fromLat, fromLng, toLat, toLng)
	mins := int(math.Ceil(km / avgSpeedKmh * 60))
	if mins < 1 {
		return 1
	}
	if mins > 180 {
		return 180
	}
	return mins
}

func haversineKm(lat1, lng1, lat2, lng2 float64) float64 {
	const earthRadiusKm = 6371.0
	toRad := func(d float64) float64 { return d * math.Pi / 180 }
	dLat := toRad(lat2 - lat1)
	dLng := toRad(lng2 - lng1)
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(toRad(lat1))*math.Cos(toRad(lat2))*math.Sin(dLng/2)*math.Sin(dLng/2)
	return 2 * earthRadiusKm * math.Asin(math.Min(1, math.Sqrt(a)))
}

// DistanceKm is the public haversine distance helper.
func DistanceKm(lat1, lng1, lat2, lng2 float64) float64 {
	return haversineKm(lat1, lng1, lat2, lng2)
}

File: apps/dashboard/utils/eta.ts
/** Approximate driving ETA from haversine distance at avgSpeedKmh (default 40). */
export function estimateEtaMinutes(
  fromLat: number,
  fromLng: number,
  toLat: number,
  toLng: number,
  avgSpeedKmh = 40
): number | null {
  if ((!fromLat && !fromLng) || (!toLat && !toLng)) return null;
  const km = haversineKm(fromLat, fromLng, toLat, toLng);
  if (!Number.isFinite(km)) return null;
  const mins = Math.ceil((km / avgSpeedKmh) * 60);
  return Math.min(180, Math.max(1, mins));
}

export function haversineKm(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const R = 6371;
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return 2 * R * Math.asin(Math.min(1, Math.sqrt(a)));
}

/** Emergency coordinates are stored as [lng, lat] strings. */
export function etaFromEmergency(
  emergency: { coordinates?: [string | number, string | number] } | null | undefined,
  requesterLat: number,
  requesterLng: number
): number | null {
  if (!emergency?.coordinates) return null;
  const lng = Number(emergency.coordinates[0]);
  const lat = Number(emergency.coordinates[1]);
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return null;
  return estimateEtaMinutes(lat, lng, requesterLat, requesterLng);
}

export function formatEta(minutes: number | null | undefined): string {
  if (minutes == null || minutes <= 0) return "—";
  if (minutes < 60) return `±${minutes} mnt`;
  const h = Math.floor(minutes / 60);
  const m = minutes % 60;
  return m ? `±${h}j ${m}m` : `±${h} jam`;
}
